In [102]:
import graphlab as gl
import matplotlib.pyplot as plt
%matplotlib inline


In [3]:
reviews = gl.load_sframe("yelp_dataset_challenge_academic_dataset/reviews/")

[INFO] graphlab.cython.cy_server: GraphLab Create v1.10 started. Logging: /tmp/graphlab_server_1465138490.log


This non-commercial license of GraphLab Create is assigned to mkkedia@dons.usfca.edu and will expire on May 31, 2017. For commercial licensing options, visit https://dato.com/buy/.


In [4]:
sm = gl.sentiment_analysis.create(reviews, 'stars', ['text'])

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 1845409

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 1797997

Number of coefficients    : 1797998

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 1         | 6        | 0.000000  | 29.821195    | 0.768852          | 0.764825            |

| 2         | 9        | 5.000000  | 47.362483    | 0.937116          | 0.915456            |

| 3         | 10       | 5.000000  | 55.046364    | 0.935499          | 0.908004            |

| 4         | 11       | 5.000000  | 62.508377    | 0.919325          | 0.890540            |

| 5         | 13       | 1.000000  | 74.389416    | 0.963755          | 0.936564            |

| 6         | 14       | 1.000000  | 82.201985    | 0.971363          | 0.943904            |

| 7         | 15       | 1.000000  | 89.730482    | 0.972742          | 0.944272            |

| 8         | 16       | 1.000000  | 97.328729    | 0.976106          | 0.944334            |

| 9         | 17       | 1.000000  | 104.901726   | 0.979091          | 0.945091            |

| 10        | 18       | 1.000000  | 112.435632   | 0.984720          | 0.945675            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

In [5]:
sm.summary()

Class                         : LogisticClassifier

Schema
------
Number of coefficients        : 1797998
Number of examples            : 1845409
Number of classes             : 2
Number of feature columns     : 1
Number of unpacked features   : 1797997

Hyperparameters
---------------
L1 penalty                    : 0.0
L2 penalty                    : 0.2

Training Summary
----------------
Solver                        : lbfgs
Solver iterations             : 10
Solver status                 : TERMINATED: Iteration limit reached.
Training time (sec)           : 115.4849

Settings
--------
Log-likelihood                : 102135.7318

Highest Positive Coefficients
-----------------------------
text["issues]                 : 22.8227
text[streaming/gaming.]       : 20.2704
text[9/5/2015.]               : 16.6593
text[erbi]                    : 15.3555
text[resurfacing!]            : 14.0152

Lowest Negative Coefficients
----------------------------
text[historical/cozy/older]   : -21.193


In [ ]:
# Get all reviews with the word burger in it
# Group by business
# Find average sentiment for those businesses
# Find top 5 
# Make sure number of reviews for the top 5 are reasonable


SyntaxError: invalid syntax (<ipython-input-10-ea3f0fc227ce>, line 1)

In [17]:
from graphlab.toolkits.feature_engineering import *

In [18]:
tokenizer = Tokenizer('text', to_lower=True, output_column_prefix='token')

In [19]:
tokenized_reviews = tokenizer.fit_transform(reviews)

In [20]:
tokenized_reviews.head()

Columns:
	business_id	str
	date	str
	review_id	str
	stars	int
	text	str
	type	str
	user_id	str
	votes	dict
	token.text	list

Rows: 10

Data:
+------------------------+------------+------------------------+-------+
|      business_id       |    date    |       review_id        | stars |
+------------------------+------------+------------------------+-------+
| 5UmKMjUEUNdYWqANhGckJw | 2012-08-01 | Ya85v4eqdd6k9Od8HbQjyA |   4   |
| 5UmKMjUEUNdYWqANhGckJw | 2014-02-13 | KPvLNJ21_4wbYNctrOwWdQ |   5   |
| 5UmKMjUEUNdYWqANhGckJw | 2015-10-31 | fFSoGV46Yxuwbr3fHNuZig |   5   |
| UsFtqoBl7naz8AVUBZMjQQ | 2013-11-08 | Di3exaUCFNw1V4kSNW5pgA |   5   |
| UsFtqoBl7naz8AVUBZMjQQ | 2014-03-29 | 0Lua2-PbqEQMjD9r89-asw |   3   |
| UsFtqoBl7naz8AVUBZMjQQ | 2014-10-29 | 7N9j5YbBHBW6qguE5DAeyA |   1   |
| UsFtqoBl7naz8AVUBZMjQQ | 2014-11-28 | mjCJR33jvUNt41iJCxDU_g |   4   |
| 3eu6MEFlq2Dg7bQh8QbdOg | 2014-02-27 | Ieh3kfZ-5J9pLju4JiQDvQ |   5   |
| 3eu6MEFlq2Dg7bQh8QbdOg | 2015-06-16 | PU28OoBSHpZLkYGCmNxlmg |   5   |
| cE27W9VPgO88Qxe4ol6y_g | 2012-08-19 | XsA6AojkWjOHA4FmuAb8XQ |   3   |
+------------------------+------------+------------------------+-------+
+-------------------------------+--------+------------------------+
|              text             |  type  |        user_id         |
+-------------------------------+--------+------------------------+
| Mr Hoagie is an institutio... | review | PUFPaY9KxDAcGqfsorJp3Q |
| Excellent food. Superb cus... | review | Iu6AxdBYGR4A0wspR9BYHA |
| Yes this place is a little... | review | auESFwWvW42h6alXgFxAXQ |
| All the food is great here... | review | uK8tzraOp4M5u3uYrqIBXg |
| We checked this place out ... | review | I_47G-R2_egp7ME5u_ltew |
| Wing sauce is like water. ... | review | PP_xoMSYlGr2pb67BbqBdA |
| Cold cheap beer. Good bar ... | review | JPPhyFE-UE453zA6K0TVgw |
| I highly recommend this pl... | review | 2d5HeDvZTDUNVog_WuUpSg |
| I am a big believer in fir... | review | BShxMIUwaJS378xcrz4Nmg |
| Decent range somewhat clos... | review | fhNxoMwwTipzjO8A9LFe8Q |
+-------------------------------+--------+------------------------+
+-------------------------------+-------------------------------+
|             votes             |           token.text          |
+-------------------------------+-------------------------------+
| {'funny': 0, 'useful': 0, ... | [mr, hoagie, is, an, insti... |
| {'funny': 0, 'useful': 0, ... | [excellent, food., superb,... |
| {'funny': 0, 'useful': 0, ... | [yes, this, place, is, a, ... |
| {'funny': 0, 'useful': 0, ... | [all, the, food, is, great... |
| {'funny': 0, 'useful': 0, ... | [we, checked, this, place,... |
| {'funny': 0, 'useful': 0, ... | [wing, sauce, is, like, wa... |
| {'funny': 0, 'useful': 0, ... | [cold, cheap, beer., good,... |
| {'funny': 0, 'useful': 0, ... | [i, highly, recommend, thi... |
| {'funny': 0, 'useful': 0, ... | [i, am, a, big, believer, ... |
| {'funny': 0, 'useful': 1, ... | [decent, range, somewhat, ... |
+-------------------------------+-------------------------------+
[10 rows x 9 columns]

In [51]:
# Selecting reviews with burgers in them
br = tokenized_reviews[tokenized_reviews['token.text'].apply(lambda x: 'burger' in set(x))]


In [58]:
print "Number of reviews which have burgers in them: {}".format(len(br))

Number of reviews which have burgers in them: 67098


In [126]:
brc = sm.classify(br)

In [127]:
# Joining the classified output with the original burger review dataset
brc['review_id'] = br['review_id']
brc = br.join(brc, on='review_id')

In [130]:
avg_pn_ratings_by_business = brc.groupby(['business_id', 'class'], {'count':gl.aggregate.COUNT('probability'),'avg':gl.aggregate.AVG('probability')})

### Summarizing the number of ratings and sentiment probabilities from Burger reviews

In [131]:
avg_pn_ratings_by_business['business_id'].sketch_summary()


+------------------+-------+----------+
|       item       | value | is exact |
+------------------+-------+----------+
|      Length      | 10527 |   Yes    |
| # Missing Values |   0   |   Yes    |
| # unique values  |  7650 |    No    |
+------------------+-------+----------+

Most frequent items:
+-------+------------------------+------------------------+------------------------+
| value | 2Mx-_HtveweqRVOyHtiSuQ | BwQiwAZTHjmoBY0yO06inQ | jqZKv-7F54guDLalXEjHxg |
+-------+------------------------+------------------------+------------------------+
| count |           3            |           3            |           3            |
+-------+------------------------+------------------------+------------------------+
+------------------------+------------------------+------------------------+------------------------+
| WaO_hAunQrZ--vI308rHQA | yJDSMugu_Pom3pJE8MRrJA | 77pQVPhksTpz-Yz8v4bK_A | 6BuCNx2gO_GYiXm2KRdMXQ |
+------------------------+------------------------+-----------------

In [134]:
# summary of counts/business of negative reviews
avg_pn_ratings_by_business[avg_pn_ratings_by_business['class']==0]['count'].sketch_summary()



+--------------------+---------------+----------+
|        item        |     value     | is exact |
+--------------------+---------------+----------+
|       Length       |      4283     |   Yes    |
|        Min         |      1.0      |   Yes    |
|        Max         |     458.0     |   Yes    |
|        Mean        | 3.83703011908 |   Yes    |
|        Sum         |    16434.0    |   Yes    |
|      Variance      | 136.978810886 |   Yes    |
| Standard Deviation | 11.7037947216 |   Yes    |
|  # Missing Values  |       0       |   Yes    |
|  # unique values   |       65      |    No    |
+--------------------+---------------+----------+

Most frequent items:
+-------+------+-----+-----+-----+-----+----+----+----+----+----+
| value |  1   |  2  |  3  |  4  |  5  | 6  | 7  | 8  | 9  | 10 |
+-------+------+-----+-----+-----+-----+----+----+----+----+----+
| count | 2182 | 737 | 392 | 200 | 167 | 98 | 83 | 70 | 44 | 41 |
+-------+------+-----+-----+-----+-----+----+----+----+----+---

In [135]:
# Summary of counts/review of positive reviews
avg_pn_ratings_by_business[avg_pn_ratings_by_business['class']==1]['count'].sketch_summary()


+--------------------+---------------+----------+
|        item        |     value     | is exact |
+--------------------+---------------+----------+
|       Length       |      6244     |   Yes    |
|        Min         |      1.0      |   Yes    |
|        Max         |     1872.0    |   Yes    |
|        Mean        | 8.11402946829 |   Yes    |
|        Sum         |    50664.0    |   Yes    |
|      Variance      | 1776.93895116 |   Yes    |
| Standard Deviation | 42.1537537018 |   Yes    |
|  # Missing Values  |       0       |   Yes    |
|  # unique values   |      141      |    No    |
+--------------------+---------------+----------+

Most frequent items:
+-------+------+------+-----+-----+-----+-----+-----+-----+----+----+
| value |  1   |  2   |  3  |  4  |  5  |  6  |  8  |  7  | 10 | 9  |
+-------+------+------+-----+-----+-----+-----+-----+-----+----+----+
| count | 2789 | 1007 | 532 | 309 | 229 | 160 | 116 | 116 | 81 | 79 |
+-------+------+------+-----+-----+-----+-----+

In [138]:

count_ratings = brc.groupby('business_id', {'count':gl.aggregate.COUNT('probability')})
count_ratings['count'].sketch_summary()


+--------------------+---------------+----------+
|        item        |     value     | is exact |
+--------------------+---------------+----------+
|       Length       |      7616     |   Yes    |
|        Min         |      1.0      |   Yes    |
|        Max         |     2330.0    |   Yes    |
|        Mean        | 8.81013655462 |   Yes    |
|        Sum         |    67098.0    |   Yes    |
|      Variance      | 2166.73180901 |   Yes    |
| Standard Deviation | 46.5481665484 |   Yes    |
|  # Missing Values  |       0       |   Yes    |
|  # unique values   |      166      |    No    |
+--------------------+---------------+----------+

Most frequent items:
+-------+------+------+-----+-----+-----+-----+-----+-----+-----+----+
| value |  1   |  2   |  3  |  4  |  5  |  6  |  7  |  8  |  9  | 10 |
+-------+------+------+-----+-----+-----+-----+-----+-----+-----+----+
| count | 3260 | 1195 | 681 | 407 | 267 | 215 | 165 | 134 | 119 | 87 |
+-------+------+------+-----+-----+-----+--

In [139]:
len(count_ratings[count_ratings['count']>31])

375

#### Notes
- There are 7650 unique businesses which have reviews of burgers
- only 375 businesses have more than 31 reviews for burgers 
- Number of positive reviews/business ranges from 1 to 1872
- Number of negative reviews/business ranges from 1 to 458 


+--------------------+---------------+----------+
|        item        |     value     | is exact |
+--------------------+---------------+----------+
|       Length       |      7616     |   Yes    |
|        Min         |      1.0      |   Yes    |
|        Max         |     2330.0    |   Yes    |
|        Mean        | 8.81013655462 |   Yes    |
|        Sum         |    67098.0    |   Yes    |
|      Variance      | 2166.73180901 |   Yes    |
| Standard Deviation | 46.5481665484 |   Yes    |
|  # Missing Values  |       0       |   Yes    |
|  # unique values   |      166      |    No    |
+--------------------+---------------+----------+

Most frequent items:
+-------+------+------+-----+-----+-----+-----+-----+-----+-----+----+
| value |  1   |  2   |  3  |  4  |  5  |  6  |  7  |  8  |  9  | 10 |
+-------+------+------+-----+-----+-----+-----+-----+-----+-----+----+
| count | 3260 | 1195 | 681 | 407 | 267 | 215 | 165 | 134 | 119 | 87 |
+-------+------+------+-----+-----+-----+--

In [118]:
businesses = gl.SFrame.read_json('yelp_dataset_challenge_academic_dataset/yelp_academic_dataset_business.json', orient='lines')

Finished parsing file /Users/mayankkedia/Documents/Jobs/Dato/yelp_dataset_challenge_academic_dataset/yelp_academic_dataset_business.json

Parsing completed. Parsed 100 lines in 1.70849 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[dict]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 59239 lines. Lines per second: 24780.7

Finished parsing file /Users/mayankkedia/Documents/Jobs/Dato/yelp_dataset_challenge_academic_dataset/yelp_academic_dataset_business.json

Parsing completed. Parsed 77445 lines in 2.70513 secs.

In [120]:
businesses[businesses['business_id']=='aGbjLWzcrnEx2ZmMCFm3EA']

Columns:
	attributes	dict
	business_id	str
	categories	list
	city	str
	full_address	str
	hours	dict
	latitude	float
	longitude	float
	name	str
	neighborhoods	list
	open	str
	review_count	int
	stars	float
	state	str
	type	str

Rows: Unknown

Data:
+-------------------------------+------------------------+
|           attributes          |      business_id       |
+-------------------------------+------------------------+
| {'Take-out': 'true', 'Acce... | aGbjLWzcrnEx2ZmMCFm3EA |
+-------------------------------+------------------------+
+------------------------+-----------+-------------------------------+
|       categories       |    city   |          full_address         |
+------------------------+-----------+-------------------------------+
| [Burgers, Restaurants] | Las Vegas | Planet Hollywood Las Vegas... |
+------------------------+-----------+-------------------------------+
+-------------------------------+---------------+----------------+
|             hours             |    latitude   |   longitude    |
+-------------------------------+---------------+----------------+
| {'Monday': {'close': '00:0... | 36.1101083959 | -115.172338486 |
+-------------------------------+---------------+----------------+
+---------------------+---------------+------+--------------+-------+-------+
|         name        | neighborhoods | open | review_count | stars | state |
+---------------------+---------------+------+--------------+-------+-------+
| Gordon Ramsay BurGR |  [The Strip]  | true |     3811     |  4.0  |   NV  |
+---------------------+---------------+------+--------------+-------+-------+
+----------+
|   type   |
+----------+
| business |
+----------+
[? rows x 15 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.